rust/src/shards/gui/layouts/disable.rs

```diff
@@ -0,0 +1,186 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright © 2022 Fragcolor Pte. Ltd. */
+
+use super::Disable;
+use crate::shard::Shard;
+use crate::shards::gui::util;
+use crate::shards::gui::BOOL_VAR_OR_NONE_SLICE;
+use crate::shards::gui::PARENTS_UI_NAME;
+use crate::types::Context;
+use crate::types::ExposedTypes;
+use crate::types::InstanceData;
+use crate::types::OptionalString;
+use crate::types::ParamVar;
+use crate::types::Parameters;
+use crate::types::ShardsVar;
+use crate::types::Type;
+use crate::types::Types;
+use crate::types::Var;
+use crate::types::ANY_TYPES;
+use crate::types::SHARDS_OR_NONE_TYPES;
+
+lazy_static! {
+  static ref DISABLE_PARAMETERS: Parameters = vec![
+    (
+      cstr!("Contents"),
+      cstr!("The UI contents."),
+      &SHARDS_OR_NONE_TYPES[..],
+    )
+      .into(),
+    (
+      cstr!("Disable"),
+      cstr!("Whether the contents should be disabled."),
+      BOOL_VAR_OR_NONE_SLICE,
+    )
+      .into(),
+  ];
+}
+
+impl Default for Disable {
+  fn default() -> Self {
+    let mut parents = ParamVar::default();
+    parents.set_name(PARENTS_UI_NAME);
+    Self {
+      parents,
+      requiring: Vec::new(),
+      contents: ShardsVar::default(),
+      disable: ParamVar::new(true.into()),
+      exposing: Vec::new(),
+    }
+  }
+}
+
+impl Shard for Disable {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.Disable")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.Disable-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "UI.Disable"
+  }
+
+  fn help(&mut self) -> OptionalString {
+    OptionalString(shccstr!("Creates a scoped child UI."))
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn inputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!(
+      "The value that will be passed to the Contents shards of the scope."
+    ))
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn outputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!("The output of this shard will be its input."))
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&DISABLE_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => self.contents.set_param(value),
+      1 => Ok(self.disable.set_param(value)),
+      _ => Err("Invalid parameter index"),
+    }
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    match index {
+      0 => self.contents.get_param(),
+      1 => self.disable.get_param(),
+      _ => Var::default(),
+    }
+  }
+
+  fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
+    self.requiring.clear();
+
+    // Add UI.Parents to the list of required variables
+    util::require_parents(&mut self.requiring, &self.parents);
+
+    Some(&self.requiring)
+  }
+
+  fn exposedVariables(&mut self) -> Option<&ExposedTypes> {
```

Comment:
Not possible, as this block is conditional, it should not expose to the upper context anything.

Reply:
It only disables the UI interaction but there are still displayed.

Reply:
Ok 👍